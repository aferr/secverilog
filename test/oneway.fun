; Domain(0)=L1, Domain(1)=L2
(declare-fun Domain (Int) Label)
(assert (= (Domain 0) L1))
(assert (= (Domain 1) L2))
