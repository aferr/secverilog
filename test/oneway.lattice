; new lattice elements L1 and L2
(declare-fun L1 () Label)
(declare-fun L2 () Label)

; lattice structure
(assert (leq L1 L2))
