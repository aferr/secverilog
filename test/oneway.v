/* 
 * This example tests a lattice with four labels: L < L1 <L2 < H.
 */
module oneway();

reg[1:0] {L} low;
reg[1:0] {H} high;
reg[1:0] {L1} d1;
reg[1:0] {L2} d2;


// domains with dependent types
// notice there are only two domains, so x, y are 1-bit long
reg {Domain x} x;
reg {Domain y} y;

// this should work, since when x is 0, label of x is Domain 0 == L1
always @(*) begin
	if (x == 0) begin
		d1 = x;
	end
end

// this should work
always @(*) begin
	if (x == 1) begin
		d2 = x;
	end
end

// this should work
always @(*) begin
	d2 = d1;
end

// this should fail
always @(*) begin
	d1 = d2;
end


endmodule
