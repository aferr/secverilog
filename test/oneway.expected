Compiling file oneway.v
Verifying file oneway.v
fail
(assert  (not(leq L2  L1)))    ; d1 = d2 @oneway.v:38
Total: 1 assertions failed
